### ISSUE TITLE
Feature request: IllegalImport rule should allow for CDATA list of classes/packages

### ISSUE THREAD

**Issue by lgoldstein** [checkstyle/checkstyle#6081]
The current [IllegalImport rule](http://checkstyle.sourceforge.net/config_imports.html#IllegalImport) configuration allows specifying "forbidden" classes and/or packages as a comma separated list of expressions (be it fully-qualified or regexp(s)) as an XML **attribute**. This imposes some limitations on the number of expressions that can be written and/or makes maintenance of the list rather cumbersome if it contains too many expressions.

The suggested feature is as follows - allow `CDATA` element instead of or in addition to the attribute - e.g.

```xml
<module name="IllegalImport">
    <illegalClasses>
        <![CDATA[
            java.text.Annotation,
            com.google.common.base.Objects,
            com.google.common.base.Predicate,
            org.apache.commons.lang3.tuple.Pair,
            ...etc....
        ]]>
    </illegalClasses>
    <illegalPkgs>
        <![CDATA[
            com.sun,
            sun,
            sunw,
           com.ibm.icu.util,
          ...etc....
        ]]>
      </illegalPkgs>
</module>
```

**Comment by romani** [comment]
> This imposes some limitations on the number of expressions that can be written and/or makes maintenance of the list rather cumbersome if it contains too many expressions.

Please provide example of your confirmation and clearly point what you want and what is inconvenient.

**Comment by lgoldstein** [comment]
>> Please provide example of your confirmation and clearly point what you want and what is inconvenient

Here is the actual example - my current configuration already has the following list:
```xml
        <module name="IllegalImport">
            <!-- Why excluded:
                 com.sun,sun,sunw - internal JVM packages with classes having many similar names to core ones
                 com.ibm.icu.util - Avoid confusion with Java core classes with similar names and functionality - e.g., Calendar
                 javassist - contains some classes with names similar to java.lang.reflect
                 antlr,groovyjarjarantlr - contains some classes with names similar to commons-lang3
                 org.apache.commons.lang - we are using the type-safe commons-lang3
                 com.querydsl.core.util - contains some classes with names similar to commons-lang3
                 org.apache.commons.collections - we are using the type-safe commons-collection4
            -->
            <property name="illegalPkgs" value="com.sun,sun,sunw,com.ibm.icu.util,javassist,antlr,groovyjarjarantlr,com.google.common,org.apache.commons.lang,com.querydsl.core.util,org.apache.commons.collections,java.awt,javax.swing,java.applet,jdk" />

            <!--
                java.text.Annotation - Not to be confused with java.lang.Annotation

                com.google.common.base.Objects - Enforce usage of java.util.Objects
                com.google.common.base.Function - Enforce usage of java.util.function.Function
                com.google.common.base.Predicate - Enforce usage of java.util.function.Predicate

                org.apache.commons.collections4.Predicate - Enforce usage of java.util.function.Predicate
                org.apache.commons.collections4.Transformer - Enforce usage of java.util.function.Function
                org.apache.commons.collections4.Closure - Enforce usage of java.util.function.Consumer

                org.apache.commons.lang3.tuple.Pair - Enforce usage of SimpleImmutableEntry

                org.springframework.util.StringUtils
                org.springframework.util.CollectionUtils
                org.springframework.cglib.core.CollectionUtils
                org.springframework.util.MapUtils - Enforce usage of Apache Commons

                org.slf4j.event.Level - Use only JUL levels
                java.text.Annotation - We only use java.lang.Annotation(s)

                org.apache.sshd.common.util.logging.AbstractLoggingBean - Enforce using our such named class
                org.apache.sshd.common.util.logging.LoggingUtils
            -->
            <property name="illegalClasses" value="java.text.Annotation,com.google.common.base.Objects,com.google.common.base.Predicate,org.apache.commons.lang3.tuple.Pair,org.apache.commons.collections4.Predicate,org.apache.commons.collections4.Transformer,org.apache.commons.collections4.Closure,org.springframework.util.StringUtils,org.springframework.util.CollectionUtils,org.springframework.util.MapUtils,org.springframework.cglib.core.CollectionUtils,org.slf4j.event.Level,org.apache.sshd.common.util.logging.AbstractLoggingBean,org.apache.sshd.common.util.logging.LoggingUtils" />
```

Note the very long list of classes and packages that are already listed - and here is where I want to go:

```
java.awt.*
javax.swing.*
com.google.*
org.apache.commons.collections.*
org.apache.commons.lang.*
javassist.*
java.applet.*
javax.sound.*
jdk.*
sun.*
net.bytebuddy.*
org.omg.*
org.apache.commons.collections.* (we want Apache Commons Collections 4.x)
org.apache.commons.lang3.tuple.Pair - we want SimpleImmutableEntry
Not to be confused with Apache Commons or our code
org.apache.sshd.common.util.io.IoUtils
org.apache.sshd.common.util.io.NoCloseInputStream
org.apache.sshd.common.util.io.NoCloseOutputStream
org.apache.sshd.common.util.io.NoCloseReader
org.apache.sshd.common.util.io.NoCloseWriter
org.apache.commons.io.input.NullInputStream
org.apache.commons.io.input.NullOutputStream
org.apache.sshd.common.util.logging.*
antlr.*
com.mchange.v2.lang.*
groovyjarjarantlr.*
com.sun*
org.joda.time.base.* (we want java.time)
org.slf4j.event.Level (we want java.util.logging.Level)
java.text.Annotation (not to be confused with java.lang.Annotation)
org.springframework.util.CollectionUtils  (we want Apache Commons)
org.springframework.util.MapUtils
org.springframework.util.StringUtils  (we want Apache Commons)
org.springframework.cglib.core.CollectionUtils
com.google.common.base.Objects (we want java.util.Objects)
com.mchange.v1.util.ListUtils
com.mchange.v1.util.SetUtils
com.querydsl.core.util.CollectionUtils
javax.xml.ws.ProtocolException (we want java.net.ProtocolException)
org.eclipse.jgit.util.StringUtils
com.google.api.client.util.StringUtils
org.apache.commons.codec.binary.StringUtils
org.sqlite.util.StringUtils
io.opencensus.*
com.univocity.parsers.common.input.EOFException (we want java.io.EOFException)
javax.print.MimeType (we want javax.activation.MimeType)
```

As can be plainly seen the list is very long (and likely to get even longer). Even if the XML standard allows me to create exceedingly long values for attributes, I am sure we can agree that maintaining such a list (adding/removing/updating/searching) is prone to errors. If there were a more "human-friendly" format, I believe it would be easier to understand and manage - I have been thinking that perhaps sub-elements is an even better solution than `CDATA`

```xml
<module name="IllegalImport">
    <illegalPkgs>
         <name>antlr</name>
         <name>com.mchange.v2.lang</name>
         <name>io.opencensus</name>
         ... etc ...
    </illegalPkgs>

    <!-- and same format for 'illegalClasses' -->
</module>
```
I truly believe such a format (in *addition* to the existing one) would really benefit such use cases

**Comment by romani** [comment]
>  I have been thinking that perhaps sub-elements is an even better solution than CDATA

DTD do not allow extra custom elements.
for fine-grained control https://checkstyle.org/config_imports.html#ImportOrder 

> Note the very long list of classes and packages that are already listed - and here is where I want to go

Now I see your problem. Property is set of string - https://checkstyle.org/config_imports.html#IllegalImport_Properties

you can make multi-line in value of XML. Example for ImportOrder - https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L273 (https://checkstyle.org/config_imports.html#ImportOrder)
It should looks pretty much the same as with CDATA.

**Comment by lgoldstein** [comment]
I did not know that...

**Comment by rnveach** [comment]
The main problem with your DTD changes is that it is super specific to this one check and property. If we went down this path, we would have to do it for every single check and we have hundreds. It would require massive changes to the DTDs all the time and custom checks may not be able to create their own without even more complex changes to the core application.

Multiple check instances will only work for specific attributes, inclusive ones. Exclusive ones won't work because all instances won't know which ones to exclude and will still print violations on them.

For comma separated properties, you can also split them up into different elements on the comma.
See https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L146-L147
````
      <property name="tokens" value="ANNOTATION_DEF"/>
      <property name="tokens" value="ANNOTATION_FIELD_DEF"/>
````
instead of
````
      <property name="tokens" value="ANNOTATION_DEF, ANNOTATION_FIELD_DEF"/>
````

**Comment by lgoldstein** [comment]
>> The main problem with your DTD changes is that it is super specific to this one check and property. If we went down this path, we would have to do it for every single check and we have hundreds of checks. It would require massive changes to the DTDs all the time when new properties are added

I am not familiar with the code, so this claim sounds a bit strange but I guess this is an internal limitation.
I do understand better the challenges and I will try to use some of the suggested solutions, although (IMO) they are incomplete. I still believe sub-elements is a good path to follow and worth considering.

**Comment by rnveach** [comment]
Its more a limitation of DTD. You must specify all allowed element names in the DTD file. Everytime we make a change to it, we up it to the next version.

**Comment by romani** [comment]
At least for now we will not change config structure
